viewer/packages/tools/src/Measurement/MeasurementTool.ts

```diff
@@ -0,0 +1,215 @@
+    this._measurements.forEach(measurement => {
+      measurement.removeMeasurement();
+    });
+    this._measurements.splice(0);
+    this._currentMeasurementIndex = -1;
+  }
+
+  /**
+   * Sets the visiblity of labels in the Measurement.
+   * @param enable
+   */
+  showMeasurementLabels(enable: boolean): void {
+    if (this._htmlOverlay) {
+      this._htmlOverlay.visible(enable);
+    }
+  }
+
+  /**
+   * Sets Measurement line width and color with @options value.
+   * @param options MeasurementOptions to set line width and color.
+   */
+  setLineOptions(options: MeasurementOptions): void {
+    // TODO 2022-07-05 larsmoa: WTF - clean up. Allowing setting options and then ignoring one of the options
+    this._options = {
+      ...MeasurementTool.defaultLineOptions,
+      distanceToLabelCallback: this._options.distanceToLabelCallback,
+      ...options
+    };
+  }
+
+  /**
+   * Dispose Measurement Tool.
+   */
+  dispose(): void {
+    this.removeAllMeasurement();
+    this._htmlOverlay.clear();
+    super.dispose();
+  }
+
+  /**
+   * Set input handling.
+   */
+  private setupEventHandling() {
+    this._viewer.on('click', this._handlePointerClick);
+  }
+
+  /**
+   * Remove input handling.
+   */
+  private removeEventHandling() {
+    this._viewer.off('click', this._handlePointerClick);
+  }
+
+  private async onPointerClick(event: any): Promise<void> {
+    const { offsetX, offsetY } = event;
+
+    const intersection = await this._viewer.getIntersectionFromPixel(offsetX, offsetY);
+
+    if (!intersection) {
+      return;
+    }
+
+    const measurementActive = this._currentMeasurementIndex !== -1;
+
+    if (!measurementActive) {
+      const camera = this._viewer.getCamera();
+      const domElement = this._viewer.domElement;
+      this._measurements.push(
+        new Measurement(domElement, camera, this._geometryGroup, this._options!, this._htmlOverlay)
+      );
+      this._currentMeasurementIndex = this._measurements.length - 1;
+      this._viewer.domElement.addEventListener('mousemove', this._handlePointerMove);
+      this._measurements[this._currentMeasurementIndex].startMeasurement(intersection.point);
+    } else {
+      this._measurements[this._currentMeasurementIndex].endMeasurement(intersection.point);
+      this._currentMeasurementIndex = -1;
+      this._viewer.domElement.removeEventListener('mousemove', this._handlePointerMove);
+    }
+    this._viewer.requestRedraw();
+  }
+
+  private onPointerMove(event: { offsetX: number; offsetY: number }) {
+    assert(this._currentMeasurementIndex !== -1);
+    this._measurements[this._currentMeasurementIndex].update(event);
+    this._viewer.requestRedraw();
+  }
+
+  /**
+   * Create and return combine ruler icon as HTMLDivElement.
+   * @returns HTMLDivElement.
+   */
+  private createCombineClusterElement() {
+    // TODO 2022-07-05 larsmoa: Move all ownership of labels here - currently responsibility is split
+    // between several classes which is *bad*
```

Comment:
Consider addressing this (and look for other TODOs in the code)

Reply:
The idea here is all measurement related objects (Line mesh, geometry, material & labels) would be in `Measurement` class, as clustering is related to tool, it would be ideal to have this here in the `MeasurementTool` class.